File: src/main/java/com/proyecto/nuclear/service/StudentService.java
package com.proyecto.nuclear.service;


import com.proyecto.nuclear.entity.Student;
import com.proyecto.nuclear.enums.EstadoPracticaEstudiante;

import java.util.List;

public interface StudentService {

    Student createStudent(Student student);

    Student updateStudent(Long id, Student student);

    Student findById(Long id);

    List<Student> findAll();

    List<Student> findByCareer(String career);

    List<Student> findByAvailability(String availability);

    List<Student> findBySkills(String skills);

    List<Student> findByCompetencies(String competencies);

    void updatePracticeStatus(
            Long studentId,
            EstadoPracticaEstudiante status);

    List<Student> filterStudents(
            String career,
            String skills,
            String competencies,
            String availability);
}

File: nuclearFront/frontend/src/app/core/services/student.service.ts
import { Injectable, inject } from '@angular/core';

import { HttpClient } from '@angular/common/http';

import { Observable } from 'rxjs';

import { Student } from '../models/student.model';

@Injectable({
  providedIn: 'root'
})
export class StudentService {

  private http = inject(HttpClient);

  private api =
    'https://proyectonuclear-production.up.railway.app/api/students';

  create(
    student: Student
  ): Observable<Student> {

    return this.http.post<Student>(
      this.api,
      student
    );
  }

  update(
    id: number,
    student: Student
  ): Observable<Student> {

    return this.http.put<Student>(
      `${this.api}/${id}`,
      student
    );
  }

  findById(
    id: number
  ): Observable<Student> {

    return this.http.get<Student>(
      `${this.api}/${id}`
    );
  }

  findAll(): Observable<Student[]> {

    return this.http.get<Student[]>(
      this.api
    );
  }

  findByCareer(
    career: string
  ): Observable<Student[]> {

    return this.http.get<Student[]>(
      `${this.api}/career/${career}`
    );
  }

  findByAvailability(
    availability: string
  ): Observable<Student[]> {

    return this.http.get<Student[]>(
      `${this.api}/availability/${availability}`
    );
  }

  findBySkills(
    skills: string
  ): Observable<Student[]> {

    return this.http.get<Student[]>(
      `${this.api}/skills/${skills}`
    );
  }

  findByCompetencies(
    competencies: string
  ): Observable<Student[]> {

    return this.http.get<Student[]>(
      `${this.api}/competencies/${competencies}`
    );
  }

  updateStatus(
    id: number,
    status: string
  ): Observable<void> {

    return this.http.put<void>(
      `${this.api}/${id}/status?status=${status}`,
      {}
    );
  }

}